packages/SwingSet/test/vat-direct.js

```diff
@@ -0,0 +1,114 @@
+import { extractMessage } from './vat-util.js';
+import { krefOf, kser, kslot, kunser } from '../src/lib/kmarshal.js';
+
+/**
+ * A testing vat that generically exposes kernel interactions
+ * which are normally handled by liveslots.
+ */
+
+/**
+ * @param {string} vpid - vat-centric promise ID p+NN
+ * @param {boolean} isRejection
+ * @param {unknown} value - fulfillment or rejection value
+ * @returns {VatOneResolution[]}
+ */
+const promiseSettlement = (vpid, isRejection, value) => [
+  [vpid, isRejection, kser(value)],
+];
+const promiseFulfillment = (vpid, value) =>
+  promiseSettlement(vpid, false, value);
+const promiseRejection = (vpid, reason) =>
+  promiseSettlement(vpid, true, reason);
+
+export default function setup(syscall, _state, _helpers, vatPowers) {
+  // eslint-disable-next-line no-unused-vars
+  const { testLog } = vatPowers;
+
+  let vatName;
+  let lastExportID = 99;
+  const incExportID = () => {
+    lastExportID += 1;
+    return lastExportID;
+  };
+  /** @type {KernelDeliveryObject[]} */
+  const dispatchLog = [];
+
+  /**
+   * Process an inbound message.
+   *
+   * @param {KernelDeliveryObject} vatDeliverObject
+   * @returns {void}
+   */
+  function dispatch(vatDeliverObject) {
+    dispatchLog.push(vatDeliverObject);
+
+    // Capture initial configuration.
+    const deliveryType = vatDeliverObject[0];
+    if (deliveryType === 'startVat') {
+      const vatParameters = kunser(vatDeliverObject[1]) || {};
+      const newVatName = vatParameters.vatName;
+      if (newVatName !== undefined) {
+        vatName = newVatName;
+      }
+    }
+
+    // React only to message deliveries.
+    if (deliveryType !== 'message') {
+      return;
+    }
+
+    const {
+      method,
+      args: argsCapdata,
+      result,
+    } = extractMessage(vatDeliverObject);
+    // Default to resolving a result promise with `undefined`.
+    /** @type {VatOneResolution[]} */
+    let resolutions = promiseFulfillment(result, undefined);
+    try {
+      // Supported methods should have names that indicate what they
+      // cause this vat to do but (for greppability) should not
+      // duplicate names that appear in actual kernel/liveslots code.
+      switch (method) {
+        case 'getDispatchLog':
+          // Return a copy of dispatchLog so the original remains mutable.
+          resolutions = promiseFulfillment(result, dispatchLog.slice());
```

Comment:
not worth changing anything, but be aware that the cool thing about setup-based test vats is that you can provide a `setup()` function that shares a lexical scope with your test code, do `t.is()` checks inside the delivery methods, record state directly into a shared closed-over array

that might simplify tests like this a bit, instead of needing to resolve a promise to the recorded data (and using `c.kpResolution` from your test case), you could just stash it somewhere. It's mostly useful if you want to ask about the vrefs that the vat is seeing: if you're interested in what the kref is, then yeah you kind of need to pass it through the dispatch/syscall boundary to let the c-list translator do it's thing

also, since you're using two vats and having one hold references into another, ok yeah you need to serialize something into a syscall.send or a resolve, and kmarshal is the easiest way to do that